File: python/src/trueform/_csg/expr.py
"""
CSG boolean expressions over operand indices.

Copyright (c) 2025 Ziga Sajovic, XLAB
Licensed for noncommercial use under the PolyForm Noncommercial License 1.0.0.
Commercial licensing available via info@polydera.com.
https://github.com/polydera/trueform
"""

_OR = -1
_AND = -2
_SUB = -3
_NOT = -4


def _as_expr(value):
    if isinstance(value, Expr):
        return value
    if isinstance(value, (int,)) and not isinstance(value, bool):
        return op(value)
    raise TypeError(
        f"expected tf.op(i) expression or operand index, got {type(value)!r}"
    )


class Expr:
    """
    A boolean expression tree over operand indices.

    Build leaves with ``tf.op(i)`` and combine with operators:
    ``|`` (union), ``&`` (intersection), ``-`` (difference),
    ``~`` (complement). Plain integers are accepted as operands:
    ``tf.op(0) - 1`` is the difference of operand 0 and operand 1.

    Examples
    --------
    >>> carved = tf.op(0) - tf.op(1) - tf.op(2)
    >>> shared = tf.op(0) & tf.op(1)
    >>> outside = ~(tf.op(0) | tf.op(1))
    """

    __slots__ = ("_program",)

    def __init__(self, program):
        self._program = program

    def program(self):
        """The flat postfix program consumed by the native layer."""
        return list(self._program)

    def _combine(self, other, code, reflected=False):
        other = _as_expr(other)
        left, right = (other, self) if reflected else (self, other)
        return Expr(left._program + right._program + [code])

    def __or__(self, other):
        return self._combine(other, _OR)

    def __ror__(self, other):
        return self._combine(other, _OR, reflected=True)

    def __and__(self, other):
        return self._combine(other, _AND)

    def __rand__(self, other):
        return self._combine(other, _AND, reflected=True)

    def __sub__(self, other):
        return self._combine(other, _SUB)

    def __rsub__(self, other):
        return self._combine(other, _SUB, reflected=True)

    def __invert__(self):
        return Expr(self._program + [_NOT])

    def __repr__(self):
        return f"tf.Expr(program={self._program})"


def op(operand_id):
    """
    A leaf expression selecting operand ``operand_id``.

    Parameters
    ----------
    operand_id : int
        Index of the operand in the ``CsgGraph``'s form list.

    Returns
    -------
    Expr
    """
    i = int(operand_id)
    if i < 0:
        raise ValueError(f"operand index must be non-negative, got {i}")
    return Expr([i])

File: typescript/src/csg/expr.ts
/*
 * Copyright (c) 2025 XLAB
 * All rights reserved.
 *
 * This file is part of trueform (trueform.polydera.com)
 *
 * Licensed for noncommercial use under the PolyForm Noncommercial
 * License 1.0.0.
 * Commercial licensing available via info@polydera.com.
 *
 * Author: Žiga Sajovic
 */

const OR = -1;
const AND = -2;
const SUB = -3;
const NOT = -4;

function asExpr(value: Expr | number): Expr {
  if (value instanceof Expr) return value;
  if (Number.isInteger(value)) return op(value);
  throw new TypeError(`expected op(i) expression or operand index, got ${value}`);
}

/**
 * A boolean expression tree over operand indices.
 *
 * Build leaves with `op(i)` and combine with the builder methods:
 * `.or()` (union), `.and()` (intersection), `.sub()` (difference),
 * `.not()` (complement). Plain integers are accepted as operands:
 * `op(0).sub(1)` is the difference of operand 0 and operand 1.
 *
 * @example
 * const carved = op(0).sub(op(1)).sub(op(2));
 * const shared = op(0).and(op(1));
 * const outside = op(0).or(op(1)).not();
 */
export class Expr {
  /** @internal */
  readonly _program: number[];

  /** @internal */
  constructor(program: number[]) {
    this._program = program;
  }

  /** The flat postfix program consumed by the native layer. */
  program(): number[] {
    return [...this._program];
  }

  /** Union with `other`. */
  or(other: Expr | number): Expr {
    return new Expr([...this._program, ...asExpr(other)._program, OR]);
  }

  /** Intersection with `other`. */
  and(other: Expr | number): Expr {
    return new Expr([...this._program, ...asExpr(other)._program, AND]);
  }

  /** Difference: this minus `other`. */
  sub(other: Expr | number): Expr {
    return new Expr([...this._program, ...asExpr(other)._program, SUB]);
  }

  /** Complement of this expression. */
  not(): Expr {
    return new Expr([...this._program, NOT]);
  }
}

/**
 * A leaf expression selecting operand `operandId` (its index in the
 * CsgGraph's form list).
 */
export function op(operandId: number): Expr {
  if (!Number.isInteger(operandId) || operandId < 0) {
    throw new RangeError(`operand index must be a non-negative integer, got ${operandId}`);
  }
  return new Expr([operandId]);
}

/** @internal */
export function programOf(expr: Expr | number): number[] {
  return asExpr(expr).program();
}
